### ISSUE TITLE
torch.no_grad, enable_grad, etc should document their behavior with forward-mode AD

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#72202]
### 📚 The doc issue

They don't modify forward-mode AD behavior. We should document it because "disabled gradient calculation" is ambiguous -- disabled reverse-mode AD gradient computation? Disabled forward-mode AD gradient computation?

Also I want to point out that detach() supports both forward-mode and reverse-mode. detach() is morally very similar to a torch.no_grad() in my mind. Is this an inconsistency?

### Suggest a potential alternative/fix

n/a

cc @brianjo @mruberry @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by albanD** [comment]
> Also I want to point out that detach() supports both forward-mode and reverse-mode. detach() is morally very similar to a torch.no_grad() in my mind. Is this an inconsistency?

It is a little bit. Originally all of these APIs where supposed to only apply to backward mode and a new set would be introduced if/when needed for no_grad. And `unpack_dual` was supposed to be the pendant to `detach()`.

In practice, because we need to handle view, having detach() being forward differentiable means that the output of detach has to keep the input alive (its base) but then we when you do things like `all_my_losses.append(loss.detach())`, you keep the whole backward graph alive!
So we had to backtrack on that idea and we made detach non-differentiable for both forward and backward AD.
